cranelift/codegen/src/remove_constant_phis.rs

```diff
@@ -0,0 +1,387 @@
+// `summaries` below.  For the `SmallVec` tuning params: most blocks have
+// few parameters, hence `4`.  And almost all blocks have either one or two
+// successors, hence `2`.
+#[derive(Debug)]
+struct BlockSummary {
+    // Formal parameters for this `Block`
+    formals: SmallVec<[Value/*Group A*/; 4]>,
+    // For each `Inst` in this block that transfers to another block: the
+    // `Inst` itself, the destination `Block`, and the actual parameters
+    // passed.  We don't bother to include transfers that pass zero parameters
+    // since that makes more work for the solver for no purpose.
+    dests: SmallVec<[(Inst, Block, SmallVec<[Value/*both Groups A and B*/; 4]>); 2]>,
+}
+impl BlockSummary {
+    fn new(formals: SmallVec<[Value; 4]>) -> Self {
+        Self {
+            formals,
+            dests: smallvec![],
+        }
+    }
+}
+
+// Solver state.  This holds a AbstractValue for each formal parameter, except
+// for those from the entry block.
+struct SolverState {
+    absvals: FxHashMap<Value /*Group A*/, AbstractValue>,
+}
+impl SolverState {
+    fn new() -> Self {
+        Self {
+            absvals: FxHashMap::default(),
+        }
+    }
+    fn get(&self, actual: Value) -> AbstractValue {
+        match self.absvals.get(&actual) {
+            Some(lp) => *lp,
+            None => panic!("SolverState::get: formal param {:?} is untracked?!", actual),
+        }
+    }
+    fn maybe_get(&self, actual: Value) -> Option<&AbstractValue> {
+        self.absvals.get(&actual)
+    }
+    fn set(&mut self, actual: Value, lp: AbstractValue) {
+        match self.absvals.insert(actual, lp) {
+            Some(_old_lp) => {}
+            None => panic!("SolverState::set: formal param {:?} is untracked?!", actual),
+        }
+    }
+}
+
+/// Detect phis in `func` that will only ever produce one value, using a
+/// classic forward dataflow analysis.  Then remove them.
+#[inline(never)]
+pub fn do_remove_constant_phis(func: &mut Function, domtree: &mut DominatorTree) {
+    let _tt = timing::remove_constant_phis();
+    debug_assert!(domtree.is_valid());
+
+    // Get the blocks, in reverse postorder
+    let mut blocks_reverse_postorder = Vec::<Block>::new();
+    for block in domtree.cfg_postorder() {
+        blocks_reverse_postorder.push(*block);
+    }
+    blocks_reverse_postorder.reverse();
+
+    // Phase 1 of 3: for each block, make a summary containing all relevant
+    // info.  The solver will iterate over the summaries, rather than having
+    // to inspect each instruction in each block.
+    let mut summaries = FxHashMap::<Block, BlockSummary>::default();
+
+    for b in &blocks_reverse_postorder {
+        let formals = func.dfg.block_params(*b);
+        let mut summary = BlockSummary::new(SmallVec::from(formals));
+
+        for inst in func.layout.block_insts(*b) {
+            let idetails = &func.dfg[inst];
+            if let BranchInfo::SingleDest(dest, _) = idetails.analyze_branch(&func.dfg.value_lists)
```

Comment:
Might be worth a comment here that multi-dest transfers (i.e., branch tables) don't carry parameters in our IR, so we only have to care about `SingleDest` here. (At least my intuition from most analyses otherwise is that one has to consider every CFG edge, which of course isn't the case here.)

Reply:
Good point.  Added.